### PR TITLE
chore(ci): try using mold in CI to speed things up

### PR DESCRIPTION
**Pull request by tobz**
Trying to see if we can bring in [`mold`](https://github.com/rui314/mold) for the purpose of speeding up CI.

In particular, we spend a _lot_ of time in the "component features" job doing linking because so many of the builds are changing a single feature flag at a time, so it's compiling a _tiny_ bit of code, reusing all the existing compiled dependencies, and then linking everything back together in the final binary.

### CODE REVIEW

File: scripts/environment/bootstrap-ubuntu-20.04.sh [hunk 1]
```diff
@@ -117,3 +117,24 @@ mv --force --verbose "$TEMP"/bin/protoc /usr/bin/protoc
...
+
+# Set Cargo to use mold as its linker.
+CARGO_OVERRIDE_DIR="${HOME}/.cargo"
+mkdir -p "${CARGO_OVERRIDE_DIR}"
+
+CARGO_OVERRIDE_CONF="${CARGO_OVERRIDE_DIR}/config.toml"
+cat <<EOF >"${CARGO_OVERRIDE_CONF}"
+[target.x86_64-unknown-linux-gnu]
+rustflags = ["-C", "linker=clang", "-C", "link-arg=-fuse-ld=/usr/bin/mold"]
+EOF
```
Comment (bruceg) [inline scripts/environment/bootstrap-ubuntu-20.04.sh:140]: What do you think about setting `rustc-wrapper` instead to something that calls `mold -run "$@"`?

**Review by bruceg — APPROVED** [review]
No objection from me, but this won't help the stated target of "component features", as `cargo check` doesn't generate code (and hence doesn't link) except for some of the dependencies.

File: scripts/environment/bootstrap-ubuntu-20.04.sh (hunk 1 above)
Comment (tobz) [reply scripts/environment/bootstrap-ubuntu-20.04.sh:140]: Hmmm... I'm not against it, but out of curiosity: is there a benefit to running that way?

**Comment by tobz** [comment]
> There are some configuration options already in the checked in .cargo/config.toml. Do we not need them in CI?

Oh, good point.  Hadn't even realized we had a repo-level configuration file for Cargo. I think we should definitely keep whatever we have in the repo there, but I believe this stuff should all be additive i.e. the user-level one and the repo-level one would be merged.

I'll double-check that, though, because it wouldn't be great if this made the CI take a different path, RUSTFLAGS-wise.

File: scripts/environment/bootstrap-ubuntu-20.04.sh (hunk 1 above)
Comment (bruceg) [reply scripts/environment/bootstrap-ubuntu-20.04.sh:140]: I recall hitting errors using mold with `rustflags`, so no real benefit if this works for you. I'll be playing around with adding `sccache` to the check-component-features run, which needs the wrapper.

**Comment by jszwedko** [comment]
@tobz Not a big deal given this was a refactoring, but given the number of changes after review on this PR, I think re-requesting it would have been prudent 🙂  I think we should limit post-review changes to small tweaks or responses to PR feedback. Any larger changes should probably trigger another review.

**Comment by tobz** [comment]
Yeah, that's a good point. I should have stepped back after getting sucked into the "try a fix, see if CI goes green" loop.  Got a little too happy that it was all green, knew I had an approval, and thought it was all good.

😓

File: benches/files.rs [hunk 2]
```diff
@@ -68,20 +68,23 @@ fn benchmark_files_no_partitions(c: &mut Criterion) {
...
 
                     (topology, input)
                 });
                 (rt, topology, input)
             },
             |(rt, topology, input)| {
                 rt.block_on(async move {
-                    let lines = random_lines(line_size).take(num_lines).map(|mut line| {
+                    let mut sink = FramedWrite::new(input, BytesCodec::new());
+                    //.sink_map_err(|e| panic!("{:?}", e));
```
Comment (blt) [inline benches/files.rs:79]: Left some commented-out code here.

File: lib/vrl/stdlib/Cargo.toml [hunk 3]
```diff
@@ -35,13 +35,13 @@ regex = { version = "1", optional = true }
...
-sha-2 = { package = "sha2", version = "0.9", optional = true }
-sha-3 = { package = "sha3", version = "0.9", optional = true }
+sha-2 = { package = "sha2", version = "0.10", optional = true }
+sha-3 = { package = "sha3", version = "0.10", optional = true }
 strip-ansi-escapes = { version = "0.1", optional = true }
 syslog_loose = { version = "0.16", optional = true }
 tracing = { version = "0.1", optional = true }
 url = { version = "2", optional = true }
-uuid = { version = "0.8", features = ["v4"], optional = true }
+uuid = { version = "1", features = ["v4"], optional = true }
```
Comment (blt) [inline lib/vrl/stdlib/Cargo.toml:44]: I think, in general, I'd prefer the use of at least "1.x" as a rule. I know we don't hold to that everywhere, but maybe we pop it in the style guide.

File: lib/vrl/stdlib/src/sha2.rs [hunk 4]
```diff
@@ -8,8 +8,8 @@ fn sha2(value: Value, variant: &Bytes) -> Resolved {
         b"SHA-256" => encode::<Sha256>(&value),
         b"SHA-384" => encode::<Sha384>(&value),
         b"SHA-512" => encode::<Sha512>(&value),
-        b"SHA-512/224" => encode::<Sha512Trunc224>(&value),
-        b"SHA-512/256" => encode::<Sha512Trunc256>(&value),
+        b"SHA-512/224" => encode::<Sha512_224>(&value),
+        b"SHA-512/256" => encode::<Sha512_256>(&value),
```
Comment (blt) [inline lib/vrl/stdlib/src/sha2.rs:12]: Crate upgrade brought in a type name change?

File: src/api/tap.rs [hunk 5]
```diff
@@ -4,7 +4,6 @@ use std::{
 };
 
 use futures::{future::try_join_all, FutureExt};
-use itertools::Itertools;
```
Comment (blt) [inline src/api/tap.rs:7]: Nice work dropping itertools where we could use the standard library instead.

Comment (tobz) [reply src/api/tap.rs:7]: Yeah, some of the stuff it does is very helpful, but uh... `.collect_vec()` vs `.collect::<Vec<_>>()` ain't enough. 😂

File: lib/vrl/stdlib/src/sha2.rs (hunk 4 above)
Comment (tobz) [reply lib/vrl/stdlib/src/sha2.rs:12]: Yeah, unfortunately.

File: lib/vrl/stdlib/Cargo.toml (hunk 3 above)
Comment (tobz) [reply lib/vrl/stdlib/Cargo.toml:44]: Yeah, that's a good point.

File: benches/files.rs (hunk 2 above)
Comment (tobz) [reply benches/files.rs:79]: 😓 

I'll make a note to clean that up.  Thanks for spotting it.

Comment (tobz) [reply benches/files.rs:79]: I'll be fixing this in https://github.com/vectordotdev/vector/pull/12572, FWIW, since I'm already working on that branch.

**Pull request merged**